zk.js/src/utxo/utxo.ts

```diff
@@ -672,113 +747,96 @@ export function createUtxo(
-    ? new Array(MERKLE_TREE_HEIGHT).fill("0")
-    : merkleProof;
-  if (
-    utxoDataHashInput !== undefined &&
-    !ownerInput &&
-    utxoDataHashInput !== "0"
-  ) {
-    throw new UtxoError(
-      UtxoErrorCode.NO_PUBLIC_KEY_PROVIDED_FOR_PROGRAM_UTXO,
-      "createUtxo",
-    );
-  }
-  const owner =
-    utxoDataHashInput !== "0" && utxoDataHashInput !== undefined
-      ? hashAndTruncateToCircuit(ownerInput!.toBytes()).toString()
-      : account.keypair.publicKey.toString();
-  const utxoDataHash =
-    utxoDataHashInput !== "0" && utxoDataHashInput !== undefined
-      ? utxoDataHashInput
-      : "0";
-  if (
-    utxoDataHashInput !== "0" &&
-    utxoDataHashInput !== undefined &&
-    !utxoData
-  ) {
-    throw new UtxoError(
-      CreateUtxoErrorCode.UTXO_DATA_UNDEFINED,
-      "createUtxo",
-      "Utxo data is undefined",
-    );
-  }
+  const { assets, amounts } = checkAssetAndAmountIntegrity(
+    uncheckedAssets,
+    uncheckedAmounts,
+  );
+
+  const { poolType, version, type } = getDefaultUtxoTypeAndVersionV0();
 
-  if (merkleTreeLeafIndexInternal === undefined) {
+  if (merkleTreeLeafIndex === undefined) {
     throw new UtxoError(
       CreateUtxoErrorCode.MERKLE_TREE_INDEX_UNDEFINED,
       "createUtxo",
       "Merkle tree index is undefined",
     );
   }
 
-  const signature = account
-    .sign(utxoHash, merkleTreeLeafIndexInternal)
-    .toString();
+  const merkleTreeLeafIndexBN = isFillingUtxo
+    ? BN_0
+    : new BN(merkleTreeLeafIndex);
+
+  const merkleProofInternal = isFillingUtxo
+    ? new Array(MERKLE_TREE_HEIGHT).fill("0")
+    : merkleProof;
+
+  const nullifier = createNullifierWithAccountSignature(
+    account,
+    hash,
+    merkleTreeLeafIndexBN,
+    lightWasm,
+  );
 
-  const nullifierInputs: NullifierInputs = {
-    signature,
-    utxoHash,
-    merkleTreeLeafIndex: merkleTreeLeafIndexInternal.toString(),
-  };
-  const nullifier = getNullifier(lightWasm, nullifierInputs);
   const utxo: Utxo = {
-    publicKey: owner,
+    owner,
     amounts,
     assets,
-    assetsCircuit: assets.map((asset, index) => {
-      if (
-        index !== 0 &&
-        asset.toBase58() === SystemProgram.programId.toBase58()
-      )
-        return "0";
-      return hashAndTruncateToCircuit(asset.toBytes()).toString();
-    }),
     blinding,
     poolType,
-    utxoHash,
-    transactionVersion,
+    hash,
+    version,
     isFillingUtxo,
     nullifier,
-    merkleTreeLeafIndex: merkleTreeLeafIndexInternal,
+    merkleTreeLeafIndex: merkleTreeLeafIndexBN.toNumber(),
     merkleProof: merkleProofInternal,
-    utxoDataHash,
-    utxoName: utxoNameInternal,
-    utxoData,
+    merkletreeId: BN_0.toNumber(), // TODO: make merkletreeId dynamic (support parallel merkle trees)
+    type,
     metaHash,
     address,
+    isPublic: false, // TODO: make isPublic dynamic
   };
   return utxo;
 }
 
+/// TODO: needs to be refactored to account for no-account access and public utxos
```

Comment:
this will need to be exposed by the wallet since you need private key material to compute the signature(hash)